File: include/apl/types/SimpleMap.java
package APL.types;

import APL.errors.SyntaxError;

public abstract class SimpleMap extends APLMap {
  @Override public Obj getRaw(Value k) {
    return getv(k.asString());
  }

  public abstract Obj getv(std::string s);
  public abstract void setv(std::string s, Obj v);

  @Override public void set(Value k, Obj v) {
    setv(k.asString(), v);
  }

  @Override public Arr allValues() {
    throw new SyntaxError("getting list of values of "+name());
  }

  @Override public Arr allKeys() {
    throw new SyntaxError("getting list of keys of "+name());
  }

  @Override public Arr kvPair() {
    throw new SyntaxError("getting entries of "+name());
  }

  @Override public int size() {
    throw new SyntaxError("getting size of "+name());
  }
}

File: APL/types/SimpleMap.cpp
#include <APL/SimpleMap.h>

namespace APL::types
{
    using SyntaxError = APL::errors::SyntaxError;

    std::shared_ptr<APL::types::Obj> SimpleMap::getRaw(std::shared_ptr<APL::types::Value> k)
    {
      return getv(k->asString());
    }

    void SimpleMap::set(std::shared_ptr<APL::types::Value> k, std::shared_ptr<APL::types::Obj> v)
    {
      setv(k->asString(), v);
    }

    std::shared_ptr<Arr> SimpleMap::allValues()
    {
      throw SyntaxError(StringHelper::wstring_to_string(L"getting list of values of " + name()));
    }

    std::shared_ptr<Arr> SimpleMap::allKeys()
    {
      throw SyntaxError(StringHelper::wstring_to_string(L"getting list of keys of " + name()));
    }

    std::shared_ptr<Arr> SimpleMap::kvPair()
    {
      throw SyntaxError(StringHelper::wstring_to_string(L"getting entries of " + name()));
    }

    int SimpleMap::size()
    {
      throw SyntaxError(StringHelper::wstring_to_string(L"getting size of " + name()));
    }
}
